core/src/main/java/org/apache/carbondata/core/carbon/datastore/BlockIndexStore.java

```diff
@@ -244,124 +222,96 @@ public static BlockIndexStore getInstance() {
    *
    * @param loadedBlockArray array of blocks which will be filled
    * @param blocksList       blocks loaded in thread
-   * @throws IndexBuilderException in case of any failure
+   * @throws CarbonUtilException in case of any failure
    */
-  private void fillLoadedBlocks(AbstractIndex[] loadedBlockArray, List<BlockInfo> blockInfos)
-      throws IndexBuilderException {
+  private void fillLoadedBlocks(AbstractIndex[] loadedBlockArray,
+      List<Future<AbstractIndex>> blocksList) throws CarbonUtilException {
+    int blockCounter = 0;
+    boolean exceptionOccurred = false;
+    Throwable exceptionRef = null;
     for (int i = 0; i < loadedBlockArray.length; i++) {
-      if (null == loadedBlockArray[i]) {
-        try {
-          loadedBlockArray[i] = mapOfBlockInfoToFuture.get(blockInfos.get(i)).get();
-        } catch (InterruptedException | ExecutionException e) {
-          throw new IndexBuilderException(e);
-        }
+      try {
+        loadedBlockArray[i] = blocksList.get(blockCounter++).get();
+      } catch (Throwable e) {
+        exceptionOccurred = true;
+        exceptionRef = e;
       }
-
+    }
+    if (exceptionOccurred) {
```

Comment:
if we move to the catch block then we will not be able decrement the access count of all the loaded 
objects